### DISCUSSION TITLE
Data(base) backup and restore

### DISCUSSION THREAD

**Discussion by jpprost** [consuldemocracy/consuldemocracy#5536] [Q&A] [answered]
Hello,

would there be a --preferably automated-- procedure to backup the content data (i.e., presumably the db), then to restore it in a fresh install?
If not, what would be the best (cleanest, easiest, etc.) option for that?

The best couple of ways I can think of are:
I/ 
(a) to manually dump the db regularly,
(b) (in case restoring is required) run the installer in a fresh environment, then restore the dump manually in the newly installed Postgres instance.

The problem is, could the manual process mess with settings, or config files that would not be adequately updated?

Alternatively:
II/
(a) dump the db
(b) the dump is restored in a new Postgres instance first, then
(c) the installer is run, assuming it can be forced to use this db instance.

But a similar question occurs with regard to the internal settings and config files.

Sorry for [cross-posting in the Consul/Community](https://community.consulproject.org/t/data-backup-and-restore/987), but I can not find the topic being addressed anywhere. Apologies if that is the case. If so, it would be very kind of you to point me in the right direction.

Cheers,
JP

**Comment by javierm** [comment, marked as answer]
Hi, @jpprost :smile:.

We currently don't provide a mechanism to automate backups. The reason is that most institutions using Consul Democracy already have their own backup systems that work for pretty much every application.

Between the methods you suggest, method I would work. Method II might work as well (we've done it in the past) but it isn't officially supported / actively tested, so there's a chance you run into some issues when the installer tries to setup the database. It will also overwrite everything you've changed in the `settings` table (the only gotcha I'm aware of, by the way).

In both cases, make sure you install the same version of Consul Democracy that you are using in the old instance.

Regarding content, you probably want to include files attached by people using the application. The full list of files and folders that might be affected are in `config/deploy.rb` (`linked_files` and `linked_dirs`). Note that not all of them need to be included in the backup, though (for example, the `tmp` folder).

Hope that helps!

**Reply by jpprost** [reply]
Hi Javier!

thank you very much for your help. 
I will test your suggestions asap and let you know. I'll tag your answer when done.
And I might as well write and share a script, to be used by tiny institutions like us ;-)

Cheers,
JP

**Reply by jpprost** [reply]
Hi javierm,

I have been working on backup and restore scripts, but I am facing a couple of issues and I am not sure this is the right place to discuss them. Should we open an issue and discuss it there?

Cheers,
JP

**Reply by javierm** [reply]
Hi, @jpprost :smile:.

Wherever you're more comfortable is OK :ok_hand:.

**Reply by jpprost** [reply]
Hi,

ok, then we may as well keep the discussion here.

I would have got a couple of questions, related to your suggestions:

@javierm> In both cases, make sure you install the same version of Consul Democracy that you are using in the old instance.


I can't find a way to automate the retrieval of the app's version number: any idea?
It would be useful in the restore process.
I considered using the git commit id, but I am not a big git expert, and since the installer is maintained in a separate repo from the main app, I don't know (a) how to retrieve the commit id of consuldemocracy/consuldemocracy that was installed, and (b) how to force the installer to use that specific commit.

@javierm> Regarding content, you probably want to include files attached by people using the application. The full list of files and folders that might be affected are in `config/deploy.rb` (`linked_files` and `linked_dirs`). Note that not all of them need to be included in the backup, though (for example, the `tmp` folder).

Ok. I make a tarball out of them, except `tmp`.
Just to make sure: I take it that all the paths in `linked_files` are relative to `~deploy/consul/current/config`, and all those in `linked_dirs` are relative to `~deploy/consul/current`?

Thanks!

Cheers,
JP

**Reply by javierm** [reply]
> I don't know (a) how to retrieve the commit id of consuldemocracy/consuldemocracy that was installed, and (b) how to force the installer to use that specific commit.

To get that commit on your development machine, run `git fetch origin --tags`. You might need to replace `origin` with whatever you call the `consuldemocracy/` remote in you `.git/config` file.

The installer and the main repository use the same version tags since version 1.1.0. So, if you'd like to install Consul Democracy 2.0.1, checkout the `2.0.1` tag **in the installer's repository**. Then, when you run the installer, it will install Consul Democracy 2.0.1. You can tell which version is going to be installed by checking the "Create first release" step in the `roles/folder_structure/tasks/main.yml` file, but note that simply changing this file to install a different version will not work; each tag of the installer supports one (and just one) Consul Democracy version.

Note that the supported operating systems are different in different versions, though. For example, installing Consul Democracy 2.0.0 (or earlier) in Ubuntu 22.04 is hard because the Ruby version used in Consul Democracy 2.0.0 doesn't support the OpenSSL version included in Ubuntu 22.04.

>  all the paths in linked_files are relative to ~deploy/consul/current/config, and all those in linked_dirs are relative to ~deploy/consul/current

Not exactly (note linked files include `config/` as well). Those files and folders are stored by default in `/home/deploy/consul/shared`. So, they would be at `/home/deploy/consul/shared/config/secrets.yml`, `/home/deploy/consul/shared/storage/`, ...

Hope that helps!

**Reply by jpprost** [reply]
Hi again @javierm :-)

thank you for the follow up! Very helpful.

> The installer and the main repository use the same version tags since version 1.1.0. 

Excellent! It should help.

> Those files and folders are stored by default in `/home/deploy/consul/shared`

Ok, good. It actually makes my life easier, no symlinks.
But then, where do you get the `.bundle` folder from? As listed in `linked_dirs`: there's a `shared/bundle` but no `shared/.bundle`, as far as I can tell.

Apart from that, I now have my bash script to backup the data. Now I move on to restoration.

Thank you!

**Reply by javierm** [reply]
> where do you get the .bundle folder from? As listed in linked_dirs: there's a shared/bundle but no shared/.bundle, as far as I can tell.

The `shared/.bundle/` folder should be there as well if you've used Capistrano to deploy a version; if you've only run the installer, it probably isn't there :thinking:. It doesn't affect much since it's only purpose is to save a few seconds every time you deploy with Capistrano, meaning that, in your case, you'd save a few seconds the first time you deploy with Capistrano, but since doing so will create that folder (if it doesn't already exist), following deployments would be fine.

The `shared/bundle` folder contains all our gem dependencies, and it's created when running `bundle install`. You don't need to backup this folder, since running the installer on the new server will also run `bundle install`.

**Reply by jpprost** [reply]
Hi @javierm,

I'm sorry to bother you again.

As I'm trying to test the restoration process, I am trying to re-install a fresh app, on a fresh remote server (an Ubuntu 24.04), from a fresh local dev host (or so I think). And here is the error I get:
```
TASK [Gathering Facts] ******************************************************************************
fatal: [92.243.26.255]: FAILED! => changed=false 
  ansible_facts: {}
  failed_modules:
    setup:
      ansible_facts:
        discovered_interpreter_python: /usr/bin/python3
      exception: |-
        Traceback (most recent call last):
          File "/home/ubuntu/.ansible/tmp/ansible-tmp-1717219521.8195047-8753136384712/AnsiballZ_setup.py", line 102, in <module>
            _ansiballz_main()
          File "/home/ubuntu/.ansible/tmp/ansible-tmp-1717219521.8195047-8753136384712/AnsiballZ_setup.py", line 94, in _ansiballz_main
            invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
          File "/home/ubuntu/.ansible/tmp/ansible-tmp-1717219521.8195047-8753136384712/AnsiballZ_setup.py", line 37, in invoke_module
            from ansible.module_utils import basic
          File "/tmp/ansible_setup_payload_9b7g7vb0/ansible_setup_payload.zip/ansible/module_utils/basic.py", line 171, in <module>
        ModuleNotFoundError: No module named 'ansible.module_utils.six.moves'
      failed: true
      module_stderr: |-
        Shared connection to 92.243.26.255 closed.
      module_stdout: |-
        Traceback (most recent call last):
          File "/home/ubuntu/.ansible/tmp/ansible-tmp-1717219521.8195047-8753136384712/AnsiballZ_setup.py", line 102, in <module>
            _ansiballz_main()
          File "/home/ubuntu/.ansible/tmp/ansible-tmp-1717219521.8195047-8753136384712/AnsiballZ_setup.py", line 94, in _ansiballz_main
            invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
          File "/home/ubuntu/.ansible/tmp/ansible-tmp-1717219521.8195047-8753136384712/AnsiballZ_setup.py", line 37, in invoke_module
            from ansible.module_utils import basic
          File "/tmp/ansible_setup_payload_9b7g7vb0/ansible_setup_payload.zip/ansible/module_utils/basic.py", line 171, in <module>
        ModuleNotFoundError: No module named 'ansible.module_utils.six.moves'
      msg: |-
        MODULE FAILURE
        See stdout/stderr for the exact error
      rc: 1
  msg: |-
    The following modules failed to execute: setup

PLAY RECAP ******************************************************************************************
92.243.26.255              : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

Here is what I did to simulate the new local and remote environments, then try to fix this issue:
- create a brand new remote server, from scratch (with a private provider)
- on the server: update apt, then upgrade all the upgradable packages
- check that python was installed
- on the local dev host: remove the old `installer/` folder
- started the install process again from scratch, including cloning the installer repo (not my own fork yet, the main one)
- on the dev host: I manually re-installed python and ansible
- I was thinking that maybe the previous install was messing with the current one, with some files being left over, so I re-booted the dev host, to make sure that the `/tmp/` folder was fresh, etc.

Now I am running out of options. Any idea?

Cheers,
JP

**Reply by jpprost** [reply]
Works fine with an ubuntu 22.04 and python3.13.

**Reply by jpprost** [reply]
HURRAY! :-) backup and restoration processes implemented and successfully tested (bash scripts).
A few drawbacks left, but nothing important.
I'll make a pull request as soon as I have got my own fork ready, if you are happy with that.

**Reply by javierm** [reply]
> Works fine with an ubuntu 22.04 and python3.13.

Which system were you using when you got the error? Just curious.

> I'll make a pull request as soon as I have got my own fork ready, if you are happy with that.

By all means :+1:. Sorry in advance if starting the review takes us a while :pray:.

**Reply by jpprost** [reply]
> Which system were you using when you got the error? Just curious.

The brand new Ubuntu 24.04 :-)

**Reply by javierm** [reply]
Good to know! As you've probably already noticed, this version isn't supported yet. It looks like adding support for it will require some changes then :relieved:.

**Reply by javierm** [reply]
Hi, @jpprost :smile:.

We've made some experiments with the installer on a Ubuntu 24.04 system and everything seems to be working fine without any changes :thinking:.

Given the error message, my best guess is that the version of Ansible you were using isn't compatible with the version of Python included in Ubuntu 24.04 :thinking:.

Unfortunately, keeping track of which version of Ansible is needed to run the installer on which system is something we haven't done so far.
